### PR TITLE
feat(retry): Offline Retry Handler

### PR DESCRIPTION
**Pull request by kelvin-lu**
### Summary

Adds a secondary retry handler, as well as a few other functionalities:
- Adds a `AsyncQueue` to `@amplitude/utils` that processes promises one by one (refactored from the transport class)
- Adds more documentation on differences between the JS and Node SDKs
- Adds an example of overriding the retry behavior
- Refactors the Transport setup to the Transport module

### Checklist

* [ ] Does your PR title have the correct [title format](https://github.com/amplitude/Amplitude-Node/blob/master/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?:

### CODE REVIEW

File: packages/node/src/transports/http.ts [hunk 1]
```diff
@@ -78,64 +73,6 @@ export class HTTPTransport implements Transport {
     return options;
   }
 
-  // Awaits the finish of all requests that have been queued up before it
```
Comment (kelvin-lu) [inline packages/node/src/transports/http.ts:81]: this all got refactored to `AsyncQueue`

File: packages/node/README.md [hunk 2]
```diff
@@ -13,5 +13,31 @@ This is Amplitude Node.js SDK written in Typescript, the 1st backend SDK for Amp
 ## Installation and Quick Start
 Please visit our :100:[Developer Center](https://developers.amplitude.com/docs/nodejs) for instructions on installing and using our the SDK.
 
+## Node and JS SDK's
+
+There might be confusion about the use case and motivations for building the Node SDK when we have a JS SDK.
+The [Amplitude-Javascript](https://github.com/amplitude/Amplitude-JavaScript) SDK is meant to be run in browser environments  and on client devices, resulting in very different designs and assumptions.
```
Comment (ricfang) [inline packages/node/README.md:19]: nit: extra space `environments  on`

File: packages/node/README.md [hunk 3]
```diff
@@ -13,5 +13,31 @@ This is Amplitude Node.js SDK written in Typescript, the 1st backend SDK for Amp
...
+For example, the Javascript contains functionality that interfaces specifically with Web API's (e.g. `referrer` and UTM attribution information, `navigator.language`), and that it can store metadata in API's such as cookies and localStorage.
```
Comment (ricfang) [inline packages/node/README.md:20]: `For example, the JavaScript SDK...`?

Comment (ricfang) [reply packages/node/README.md:20]: `API's` should there be an apostrophe?

File: packages/node/README.md [hunk 4]
```diff
@@ -13,5 +13,31 @@ This is Amplitude Node.js SDK written in Typescript, the 1st backend SDK for Amp
...
+It also makes other assumptions, such as that there is only one active "identity" sending events at a time.
+
+In contrast, the Node SDK is meant to be run in a Node environment, and kept as flexible and modular as possible as to accomodate the various different use cases it may be used in.
```
Comment (ricfang) [inline packages/node/README.md:23]: `accommodate`

File: packages/node/src/retry/offlineRetry.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  private _onSendSuccess(): void {
+    // If we successfully sent all the events, reset the queue!
+    this._eventsToRetry = [];
+  }
+
+  private _onSendError(events: readonly Event[], response: Response): void {
+    let newEventsToRetry: Event[] = Array.from(events);
+    // See if there are any events we can immediately throw out
+    if (response.status === Status.Invalid) {
+      if (typeof response.body?.missingField === 'string' || events.length === 1) {
```
Comment (ricfang) [inline packages/node/src/retry/offlineRetry.ts:66]: does `missingField` only get returned if the entire payload is invalid? why not always do the logic in the `else if`?

File: packages/node/src/retry/offlineRetry.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        // Don't retry anything if there's an issue with the entire payload
+        // or if there's only one event and its invalid
+        newEventsToRetry = [];
+      } else if (response.body !== undefined) {
+        const invalidEventIndices = new Set<number>(collectInvalidEventIndices(response));
+        newEventsToRetry = events.filter((_, index) => !invalidEventIndices.has(index));
+      }
+    } else if (response.status === Status.Success) {
```
Comment (ricfang) [inline packages/node/src/retry/offlineRetry.ts:74]: will this ever be `true`?

File: packages/utils/src/queue.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
+interface QueueObject {
```
Comment (ricfang) [inline packages/utils/src/queue.ts:1]: this looks the same as the other file so I didn't review too closely

File: packages/node/src/retry/offlineRetry.ts (hunk 5 above)
Comment (kelvin-lu) [reply packages/node/src/retry/offlineRetry.ts:66]: right! `missingField` is if there's a field missing on the ENTIRE payload. There's very little chance of this happening besides not giving an API key at all, but thought to include this case.

File: packages/node/src/retry/offlineRetry.ts (hunk 6 above)
Comment (kelvin-lu) [reply packages/node/src/retry/offlineRetry.ts:74]: I can't imagine it so far, but life continues to find new ways to surprise us so i left this in defensively :)

**Review by jooohhn — APPROVED** [review]
LGTM!

**Pull request merged**